### PR TITLE
support waitUntil option for `page.setContent`

### LINKED ISSUE #784: Support waitUntil option for `page.setContent`
https://github.com/kblok/puppeteer/pull/14/commits/927d0f443b832609c91ecd7bdbf1945a5d219c60

### PR DESCRIPTION
**Pull request by Meir017**
closes #784

### CODE REVIEW

File: lib/PuppeteerSharp.Tests/PageTests/SetContentTests.cs [hunk 1]
```diff
@@ -44,5 +44,20 @@ public async Task ShouldWorkWithHtml4Doctype()
...
+        [Fact]
+        public async Task ShouldAwaitResourceToLoad()
+        {
+            var imgPath = "/img.png";
+            var imgResponse = new TaskCompletionSource<bool>();
+            Server.SetRoute(imgPath, context => imgResponse.Task);
+            var loaded = false;
+            var contentTask = Page.SetContentAsync($"<img src=\"${TestConstants.ServerUrl + imgPath}\"></img>")
+                .ContinueWith(_ => loaded = true);
+            await Server.WaitForRequest(imgPath);
```
Comment (kblok) [inline lib/PuppeteerSharp.Tests/PageTests/SetContentTests.cs]: We might have some race conditions here. I'd follow this pattern.

File: lib/PuppeteerSharp/Frame.cs [hunk 2]
```diff
@@ -544,21 +544,40 @@ public Task<string> GetContentAsync()
...
+        /// <param name="options">The options</param>
         /// <returns>Task.</returns>
-        /// <seealso cref="Page.SetContentAsync(string)"/>
-        public Task SetContentAsync(string html)
-            => EvaluateFunctionAsync(@"html => {
+        /// <seealso cref="Page.SetContentAsync(string, NavigationOptions)"/>
+        public async Task SetContentAsync(string html, NavigationOptions options = null)
+        {
+            var waitUntil = options?.WaitUntil ?? new[] { WaitUntilNavigation.Load };
+            var timeout = options?.Timeout ?? 30_000;
```
Comment (kblok) [inline lib/PuppeteerSharp/Frame.cs]: Let's move `30_000` to a const.

File: lib/PuppeteerSharp/Frame.cs [hunk 3]
```diff
@@ -544,21 +544,40 @@ public Task<string> GetContentAsync()
...
-            }", html);
-
+            }", html);
+
+            var watcher = new LifecycleWatcher(FrameManager, this, timeout, options);
+
+            await Task.WhenAny(
+                watcher.TimeoutOrTerminationTask,
+                watcher.LifecycleTask).ConfigureAwait(false);
+            if (watcher.TimeoutOrTerminationTask.IsCompleted && watcher.TimeoutOrTerminationTask.Result.IsFaulted)
```
Comment (kblok) [inline lib/PuppeteerSharp/Frame.cs]: We could change this to
```cs
var task = await Task.WhenAny(
     watcher.TimeoutOrTerminationTask,
     watcher.LifecycleTask).ConfigureAwait(false);
await task;          
```

if `task` is `TimeoutOrTerminationTask` we will get the TimeoutException.

File: lib/PuppeteerSharp.Tests/PageTests/SetContentTests.cs (hunk 1 above)
Comment (Meir017) [reply lib/PuppeteerSharp.Tests/PageTests/SetContentTests.cs]: what pattern?

File: lib/PuppeteerSharp/Frame.cs (hunk 2 above)
Comment (Meir017) [reply lib/PuppeteerSharp/Frame.cs]: we have a few places where we declare a default timeout of 30000
in `Page` & `FrameManager` and a few of the options classes
where should this constant be added to?

File: lib/PuppeteerSharp.Tests/PageTests/SetContentTests.cs (hunk 1 above)
Comment (Meir017) [reply lib/PuppeteerSharp.Tests/PageTests/SetContentTests.cs]: not sure what to do about this test 😢

**Comment by kblok** [comment]
Sorry man, I was out for a few days, would you mind checking this commit https://github.com/kblok/puppeteer-sharp/commit/1856b4a4f98b969dd4d3c7f322542afb0be35f18 ?

**Comment by kblok** [comment]
We should be green with this @Meir017  :)
https://github.com/kblok/puppeteer-sharp/commit/f463c479e20d5a5c5020fa95b577f4041c2d4d95

**Pull request merged**